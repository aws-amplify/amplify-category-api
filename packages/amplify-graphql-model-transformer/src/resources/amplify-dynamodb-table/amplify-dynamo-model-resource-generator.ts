import * as cdk from 'aws-cdk-lib';
import { TransformerContextProvider } from '@aws-amplify/graphql-transformer-interfaces';
import { ModelResourceIDs, ResourceConstants } from 'graphql-transformer-common';
import { ObjectTypeDefinitionNode } from 'graphql';
import { setResourceName } from '@aws-amplify/graphql-transformer-core';
import { AttributeType, StreamViewType, TableEncryption } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

import { Duration, aws_iam, aws_lambda, custom_resources, aws_logs } from 'aws-cdk-lib';
import { DynamoModelResourceGenerator } from '../dynamo-model-resource-generator';
import * as path from 'path';
import { AmplifyDynamoDBTable } from './amplify-dynamodb-table-construct';

/**
 * AmplifyDynamoModelResourceGenerator is a subclass of DynamoModelResourceGenerator,
 * provisioning the DynamoDB tables with the custom resource instead of pre-defined DynamoDB table CFN template
 */

export const ITERATIVE_TABLE_STACK_NAME = 'AmplifyTableManager';
export class AmplifyDynamoModelResourceGenerator extends DynamoModelResourceGenerator {
  private customResourceServiceToken: string = '';
  private ddbManagerPolicy?: aws_iam.Policy;

  generateResources(ctx: TransformerContextProvider): void {
    if (!this.isEnabled()) {
      return;
    }

    if (this.isProvisioned()) {
      // add model related-parameters to the root stack
      const rootStack = cdk.Stack.of(ctx.stackManager.scope);
      this.createDynamoDBParameters(rootStack, false);

      const tableManagerStack = ctx.stackManager.getScopeFor('AmplifyTableCustomProvider', ITERATIVE_TABLE_STACK_NAME);
      this.createCustomProviderResource(tableManagerStack, ctx);
    }

    this.models.forEach((model) => {
      // This name is used by the mock functionality. Changing this can break mock.
      const tableBaseName = ctx.resourceHelper.getModelNameMapping(model!.name.value);
      const tableLogicalName = ModelResourceIDs.ModelTableResourceID(tableBaseName);
      const scope = ctx.stackManager.getScopeFor(tableLogicalName, tableBaseName);

      this.createModelTable(scope, model, ctx);
    });

    if (this.ddbManagerPolicy) {
      this.ddbManagerPolicy?.addStatements(
        new aws_iam.PolicyStatement({
          actions: [
            'dynamodb:CreateTable',
            'dynamodb:UpdateTable',
            'dynamodb:DeleteTable',
            'dynamodb:DescribeTable',
            'dynamodb:DescribeContinuousBackups',
            'dynamodb:DescribeTimeToLive',
            'dynamodb:UpdateContinuousBackups',
            'dynamodb:UpdateTimeToLive',
          ],
          resources: [
            cdk.Fn.sub('arn:aws:dynamodb:${AWS::Region}:${AWS::AccountId}:table/*-${apiId}-${envName}', {
              apiId: ctx.api.apiId,
              envName: ctx.synthParameters.amplifyEnvironmentName,
            }),
          ],
        }),
      );
    }

    this.generateResolvers(ctx);
  }

  protected createCustomProviderResource(scope: Construct, context: TransformerContextProvider): void {
    // Policy that grants access to Create/Update/Delete DynamoDB tables
    this.ddbManagerPolicy = new aws_iam.Policy(scope, 'CreateUpdateDeleteTablesPolicy');

    const lambdaCode = aws_lambda.Code.fromAsset(
      path.join(__dirname, '..', '..', '..', 'lib', 'resources', 'amplify-dynamodb-table', 'amplify-table-manager-lambda'),
    );

    // lambda that will handle DDB CFN events
    const gsiOnEventHandler = new aws_lambda.Function(scope, ResourceConstants.RESOURCES.TableManagerOnEventHandlerLogicalID, {
      runtime: aws_lambda.Runtime.NODEJS_18_X,
      code: lambdaCode,
      handler: 'amplify-table-manager-handler.onEvent',
      timeout: Duration.minutes(14),
    });

    // lambda that will poll for provisioning to complete
    const gsiIsCompleteHandler = new aws_lambda.Function(scope, ResourceConstants.RESOURCES.TableManagerIsCompleteHandlerLogicalID, {
      runtime: aws_lambda.Runtime.NODEJS_18_X,
      code: lambdaCode,
      handler: 'amplify-table-manager-handler.isComplete',
      timeout: Duration.minutes(14),
    });

    this.ddbManagerPolicy.attachToRole(gsiOnEventHandler.role!);
    this.ddbManagerPolicy.attachToRole(gsiIsCompleteHandler.role!);
    const customResourceProvider = new custom_resources.Provider(scope, ResourceConstants.RESOURCES.TableManagerCustomProviderLogicalID, {
      onEventHandler: gsiOnEventHandler,
      isCompleteHandler: gsiIsCompleteHandler,
      logRetention: aws_logs.RetentionDays.ONE_MONTH,
      queryInterval: Duration.seconds(30),
      totalTimeout: Duration.hours(2),
    });
    this.customResourceServiceToken = customResourceProvider.serviceToken;
  }

  protected createModelTable(scope: Construct, def: ObjectTypeDefinitionNode, context: TransformerContextProvider): void {
    const modelName = def!.name.value;
    const tableLogicalName = ModelResourceIDs.ModelTableResourceID(modelName);
    const tableName = context.resourceHelper.generateTableName(modelName);

    // Add parameters.
    const { readIops, writeIops, billingMode, pointInTimeRecovery, enableSSE } = this.createDynamoDBParameters(scope, true);

    // Add conditions.
    new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.HasEnvironmentParameter, {
      expression: cdk.Fn.conditionNot(cdk.Fn.conditionEquals(context.synthParameters.amplifyEnvironmentName, ResourceConstants.NONE)),
    });
    const useSSE = new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.ShouldUseServerSideEncryption, {
      expression: cdk.Fn.conditionEquals(enableSSE, 'true'),
    });
    const usePayPerRequestBilling = new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.ShouldUsePayPerRequestBilling, {
      expression: cdk.Fn.conditionEquals(billingMode, 'PAY_PER_REQUEST'),
    });
    const usePointInTimeRecovery = new cdk.CfnCondition(scope, ResourceConstants.CONDITIONS.ShouldUsePointInTimeRecovery, {
      expression: cdk.Fn.conditionEquals(pointInTimeRecovery, 'true'),
    });

    const removalPolicy = this.options.EnableDeletionProtection ? cdk.RemovalPolicy.RETAIN : cdk.RemovalPolicy.DESTROY;

    // TODO: The attribute of encryption and TTL should be added
    const table = new AmplifyDynamoDBTable(scope, `${tableLogicalName}`, {
      customResourceServiceToken: this.customResourceServiceToken,
      allowDestructiveGraphqlSchemaUpdates: context.transformParameters.allowDestructiveGraphqlSchemaUpdates,
      replaceTableUponGsiUpdate: context.transformParameters.replaceTableUponGsiUpdate,
      tableName,
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING,
      },
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
      encryption: TableEncryption.DEFAULT,
      removalPolicy,
      ...(context.isProjectUsingDataStore() ? { timeToLiveAttribute: '_ttl' } : undefined),
    });
    setResourceName(table, { name: modelName, setOnDefaultChild: false });

    // construct a wrapper around the custom table to allow normal CDK operations on top of it
    const tableRepresentative = table.tableFromAttr;
    setResourceName(tableRepresentative, { name: modelName, setOnDefaultChild: false });

    const cfnTable = table.node.defaultChild?.node.defaultChild as cdk.CfnCustomResource;
    setResourceName(cfnTable, { name: modelName, setOnDefaultChild: false });
    cfnTable.addPropertyOverride(
      'provisionedThroughput',
      cdk.Fn.conditionIf(usePayPerRequestBilling.logicalId, cdk.Fn.ref('AWS::NoValue'), {
        ReadCapacityUnits: readIops,
        WriteCapacityUnits: writeIops,
      }),
    );
    cfnTable.addPropertyOverride(
      'pointInTimeRecoverySpecification',
      cdk.Fn.conditionIf(usePointInTimeRecovery.logicalId, { PointInTimeRecoveryEnabled: true }, cdk.Fn.ref('AWS::NoValue')),
    );
    cfnTable.addPropertyOverride(
      'billingMode',
      cdk.Fn.conditionIf(usePayPerRequestBilling.logicalId, 'PAY_PER_REQUEST', cdk.Fn.ref('AWS::NoValue')).toString(),
    );
    cfnTable.addPropertyOverride('sseSpecification', {
      sseEnabled: cdk.Fn.conditionIf(useSSE.logicalId, true, false),
    });

    const streamArnOutputId = `GetAtt${ModelResourceIDs.ModelTableStreamArn(def!.name.value)}`;
    if (table.tableStreamArn) {
      new cdk.CfnOutput(cdk.Stack.of(scope), streamArnOutputId, {
        value: table.tableStreamArn,
        description: 'Your DynamoDB table StreamArn.',
        exportName: cdk.Fn.join(':', [context.api.apiId, 'GetAtt', tableLogicalName, 'StreamArn']),
      });
    }

    const tableNameOutputId = `GetAtt${tableLogicalName}Name`;
    new cdk.CfnOutput(cdk.Stack.of(scope), tableNameOutputId, {
      value: table.tableName,
      description: 'Your DynamoDB table name.',
      exportName: cdk.Fn.join(':', [context.api.apiId, 'GetAtt', tableLogicalName, 'Name']),
    });

    const role = this.createIAMRole(context, def, scope, tableName);
    const tableDataSourceLogicalName = `${def!.name.value}Table`;
    this.createModelTableDataSource(def, context, tableRepresentative, scope, role, tableDataSourceLogicalName);
  }
}
