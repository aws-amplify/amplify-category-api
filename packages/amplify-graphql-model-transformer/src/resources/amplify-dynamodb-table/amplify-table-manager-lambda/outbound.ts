// The contents of this file were taken from the AWS CDK provider framework.
// https://github.com/aws/aws-cdk/blob/c52ff08/packages/aws-cdk-lib/custom-resources/lib/provider-framework/runtime/outbound.ts

/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable prefer-const */
/* eslint-disable prefer-arrow/prefer-arrow-functions */
/* eslint-disable func-style */
import * as https from 'https';
import { SFN, StartExecutionInput, StartExecutionOutput } from '@aws-sdk/client-sfn';

const FRAMEWORK_HANDLER_TIMEOUT = 900000; // 15 minutes

// In order to honor the overall maximum timeout set for the target process,
// the default 2 minutes from AWS SDK has to be overriden:
// https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#httpOptions-property
const awsSdkConfig = {
  httpOptions: { timeout: FRAMEWORK_HANDLER_TIMEOUT },
};

async function defaultHttpRequest(options: https.RequestOptions, responseBody: string) {
  return new Promise((resolve, reject) => {
    try {
      const request = https.request(options, resolve);
      request.on('error', reject);
      request.write(responseBody);
      request.end();
    } catch (e) {
      reject(e);
    }
  });
}

let sfn: SFN;

async function defaultStartExecution(req: StartExecutionInput): Promise<StartExecutionOutput> {
  if (!sfn) {
    sfn = new SFN(awsSdkConfig);
  }

  return sfn.startExecution(req);
}

export let startExecution = defaultStartExecution;
export let httpRequest = defaultHttpRequest;
