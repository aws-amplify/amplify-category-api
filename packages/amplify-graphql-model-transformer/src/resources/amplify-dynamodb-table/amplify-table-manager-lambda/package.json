{
  "name": "amplify-table-manager-lambda",
  "version": "1.0.0",
  "description": "Amplify Table Manager Lambda",
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "private": true,
  "dependencies": {
    "lodash.isequal": "^4.5.0"
  }
}
