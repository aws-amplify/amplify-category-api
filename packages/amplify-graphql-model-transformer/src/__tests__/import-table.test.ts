import { TableDescription } from '@aws-sdk/client-dynamodb';
import {
  getExpectedTableProperties,
  getImportedTableComparisonProperties,
  validateImportedTableProperties,
  TableComparisonProperties,
} from '../resources/amplify-dynamodb-table/amplify-table-manager-lambda/import-table';
import {
  extractTableInputFromEvent,
  toCreateTableInput,
} from '../resources/amplify-dynamodb-table/amplify-table-manager-lambda/amplify-table-manager-handler';
import * as ddbTableManagerLambda from '../resources/amplify-dynamodb-table/amplify-table-manager-lambda/amplify-table-manager-handler';
import { RequestType } from '../resources/amplify-dynamodb-table/amplify-table-manager-lambda-types';

jest.spyOn(ddbTableManagerLambda, 'getLambdaTags').mockReturnValue(
  Promise.resolve([
    { Key: 'key1', Value: 'value1' },
    { Key: 'key2', Value: 'value2' },
    { Key: 'key3', Value: 'value3' },
  ]),
);

describe('import-table', () => {
  test('getExpectedTableProperties', async () => {
    const mockEvent = {
      ServiceToken: 'mockServiceToken',
      ResponseURL: 'mockResponseURL',
      StackId: 'mockStackId',
      RequestId: 'mockRequestId',
      LogicalResourceId: 'mockLogicalResourceId',
      ResourceType: 'mockResourceType',
      RequestType: 'Create' as RequestType,
      ResourceProperties: {
        ServiceToken: 'mockServiceToken',
        tableName: 'mockTableName',
        attributeDefinitions: [
          {
            attributeName: 'todoId',
            attributeType: 'S',
          },
          {
            attributeName: 'name',
            attributeType: 'S',
          },
          {
            attributeName: 'name2',
            attributeType: 'S',
          },
        ],
        keySchema: [
          {
            attributeName: 'todoId',
            keyType: 'HASH',
          },
          {
            attributeName: 'name',
            keyType: 'RANGE',
          },
        ],
        globalSecondaryIndexes: [
          {
            indexName: 'byName2',
            keySchema: [
              {
                attributeName: 'name2',
                keyType: 'HASH',
              },
            ],
            projection: {
              projectionType: 'ALL',
            },
            provisionedThroughput: {
              readCapacityUnits: '5',
              writeCapacityUnits: '5',
            },
          },
        ],
        billingMode: 'PROVISIONED',
        provisionedThroughput: {
          readCapacityUnits: '5',
          writeCapacityUnits: '5',
        },
        sseSpecification: {
          sseEnabled: 'true',
        },
        streamSpecification: {
          streamViewType: 'NEW_AND_OLD_IMAGES',
        },
      },
    };

    const mockContext = {
      invokedFunctionArn: 'mockFunctionArn',
    };
    const tableDef = await extractTableInputFromEvent(mockEvent, mockContext);
    const createTableInput = toCreateTableInput(tableDef);
    const expectedTableProperties = getExpectedTableProperties(createTableInput);
    expect(expectedTableProperties).toMatchSnapshot();
  });

  describe('validateImportedTableProperties', () => {
    describe('matching properties', () => {
      test('AttributeDefinitions', () => {
        const actual: TableDescription = {
          AttributeDefinitions: [
            {
              AttributeName: 'todoId',
              AttributeType: 'S',
            },
            {
              AttributeName: 'name',
              AttributeType: 'S',
            },
            {
              AttributeName: 'name2',
              AttributeType: 'S',
              // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
              foo: 'bar',
              baz: undefined,
            },
          ],
        };
        const expected: TableComparisonProperties = {
          AttributeDefinitions: [
            {
              AttributeName: 'todoId',
              AttributeType: 'S',
            },
            {
              AttributeName: 'name',
              AttributeType: 'S',
            },
            {
              AttributeName: 'name2',
              AttributeType: 'S',
            },
          ],
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });

      test('KeySchema', () => {
        const actual: TableDescription = {
          KeySchema: [
            {
              AttributeName: 'todoId',
              KeyType: 'HASH',
            },
            {
              AttributeName: 'name',
              KeyType: 'RANGE',
              // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
              foo: 'bar',
              baz: undefined,
            },
          ],
        };
        const expected: TableComparisonProperties = {
          KeySchema: [
            {
              AttributeName: 'todoId',
              KeyType: 'HASH',
            },
            {
              AttributeName: 'name',
              KeyType: 'RANGE',
            },
          ],
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });

      test('GlobalSecondaryIndexes', () => {
        const actual: TableDescription = {
          GlobalSecondaryIndexes: [
            {
              IndexName: 'byName2',
              KeySchema: [
                {
                  AttributeName: 'name2',
                  KeyType: 'HASH',
                },
              ],
              Projection: {
                ProjectionType: 'ALL',
              },
              ProvisionedThroughput: {
                ReadCapacityUnits: 5,
                WriteCapacityUnits: 5,
              },
              // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
              foo: 'bar',
              baz: undefined,
            },
          ],
        };
        const expected: TableComparisonProperties = {
          GlobalSecondaryIndexes: [
            {
              IndexName: 'byName2',
              KeySchema: [
                {
                  AttributeName: 'name2',
                  KeyType: 'HASH',
                },
              ],
              Projection: {
                ProjectionType: 'ALL',
              },
              ProvisionedThroughput: {
                ReadCapacityUnits: 5,
                WriteCapacityUnits: 5,
              },
            },
          ],
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });

      test('BillingModeSummary', () => {
        const actual: TableDescription = {
          BillingModeSummary: {
            BillingMode: 'PROVISIONED',
            // should be ignored
            LastUpdateToPayPerRequestDateTime: new Date(),
            // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
            foo: 'bar',
            baz: undefined,
          },
        };
        const expected: TableComparisonProperties = {
          BillingModeSummary: {
            BillingMode: 'PROVISIONED',
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });

      test('ProvisionedThroughput', () => {
        const actual: TableDescription = {
          ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 5,
            // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
            foo: 'bar',
            baz: undefined,
            // should be ingnored
            LastDecreaseDateTime: new Date(),
            LastIncreaseDateTime: new Date(),
            NumberOfDecreasesToday: 0,
          },
        };
        const expected: TableComparisonProperties = {
          ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 5,
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });

      test('StreamSpecification', () => {
        const actual: TableDescription = {
          StreamSpecification: {
            StreamEnabled: true,
            StreamViewType: 'NEW_AND_OLD_IMAGES',
            // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
            foo: 'bar',
            baz: undefined,
          },
        };
        const expected: TableComparisonProperties = {
          StreamSpecification: {
            StreamEnabled: true,
            StreamViewType: 'NEW_AND_OLD_IMAGES',
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });

      test('SSEDescription', () => {
        const actual: TableDescription = {
          SSEDescription: {
            SSEType: 'KMS',
            // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
            foo: 'bar',
            baz: undefined,
            // should be ignored
            Status: 'ENABLED',
          },
        };
        const expected: TableComparisonProperties = {
          SSEDescription: {
            SSEType: 'KMS',
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });

      test('DeletionProtectionEnabled', () => {
        const actual: TableDescription = {
          DeletionProtectionEnabled: true,
        };
        const expected: TableComparisonProperties = {
          DeletionProtectionEnabled: true,
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected)).not.toThrow();
      });
    });

    describe('non-matching properties', () => {
      test('AttributeDefinitions', () => {
        const actual: TableDescription = {
          AttributeDefinitions: [
            {
              AttributeName: 'todoId',
              AttributeType: 'S',
              // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
              foo: 'bar',
              baz: undefined,
            },
            {
              AttributeName: 'name',
              AttributeType: 'S',
            },
            {
              AttributeName: 'name2',
              AttributeType: 'S',
            },
          ],
        };
        const expected: TableComparisonProperties = {
          AttributeDefinitions: [
            {
              AttributeName: 'todoId',
              AttributeType: 'S',
            },
            {
              AttributeName: 'differentName',
              AttributeType: 'S',
            },
            {
              AttributeName: 'name2',
              AttributeType: 'S',
            },
          ],
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          AttributeDefinitions does not match the expected value.
          Actual: [{\\"AttributeName\\":\\"todoId\\",\\"AttributeType\\":\\"S\\"},{\\"AttributeName\\":\\"name\\",\\"AttributeType\\":\\"S\\"},{\\"AttributeName\\":\\"name2\\",\\"AttributeType\\":\\"S\\"}]
          Expected: [{\\"AttributeName\\":\\"todoId\\",\\"AttributeType\\":\\"S\\"},{\\"AttributeName\\":\\"differentName\\",\\"AttributeType\\":\\"S\\"},{\\"AttributeName\\":\\"name2\\",\\"AttributeType\\":\\"S\\"}]"
        `);
      });

      test('KeySchema', () => {
        const actual: TableDescription = {
          KeySchema: [
            {
              AttributeName: 'todoId',
              KeyType: 'HASH',
              // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
              foo: 'bar',
              baz: undefined,
            },
            {
              AttributeName: 'name',
              KeyType: 'RANGE',
            },
          ],
        };
        const expected: TableComparisonProperties = {
          KeySchema: [
            {
              AttributeName: 'todoId',
              KeyType: 'HASH',
            },
            {
              AttributeName: 'differentName',
              KeyType: 'RANGE',
            },
          ],
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          KeySchema does not match the expected value.
          Actual: [{\\"AttributeName\\":\\"todoId\\",\\"KeyType\\":\\"HASH\\"},{\\"AttributeName\\":\\"name\\",\\"KeyType\\":\\"RANGE\\"}]
          Expected: [{\\"AttributeName\\":\\"todoId\\",\\"KeyType\\":\\"HASH\\"},{\\"AttributeName\\":\\"differentName\\",\\"KeyType\\":\\"RANGE\\"}]"
        `);
      });

      test('GlobalSecondaryIndexes', () => {
        const actual: TableDescription = {
          GlobalSecondaryIndexes: [
            {
              IndexName: 'byName2',
              KeySchema: [
                {
                  AttributeName: 'name2',
                  KeyType: 'HASH',
                },
              ],
              Projection: {
                ProjectionType: 'ALL',
              },
              ProvisionedThroughput: {
                ReadCapacityUnits: 5,
                WriteCapacityUnits: 5,
              },
              // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
              foo: 'bar',
              baz: undefined,
            },
          ],
        };
        const expected: TableComparisonProperties = {
          GlobalSecondaryIndexes: [
            {
              IndexName: 'byName2',
              KeySchema: [
                {
                  AttributeName: 'name2',
                  KeyType: 'HASH',
                },
              ],
              Projection: {
                ProjectionType: 'ALL',
              },
              ProvisionedThroughput: {
                ReadCapacityUnits: 5,
                WriteCapacityUnits: 10,
              },
            },
          ],
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          GlobalSecondaryIndexes does not match the expected value.
          Actual: [{\\"IndexName\\":\\"byName2\\",\\"KeySchema\\":[{\\"AttributeName\\":\\"name2\\",\\"KeyType\\":\\"HASH\\"}],\\"Projection\\":{\\"ProjectionType\\":\\"ALL\\"},\\"ProvisionedThroughput\\":{\\"ReadCapacityUnits\\":5,\\"WriteCapacityUnits\\":5}}]
          Expected: [{\\"IndexName\\":\\"byName2\\",\\"KeySchema\\":[{\\"AttributeName\\":\\"name2\\",\\"KeyType\\":\\"HASH\\"}],\\"Projection\\":{\\"ProjectionType\\":\\"ALL\\"},\\"ProvisionedThroughput\\":{\\"ReadCapacityUnits\\":5,\\"WriteCapacityUnits\\":10}}]"
        `);
      });

      test('BillingModeSummary', () => {
        const actual: TableDescription = {
          BillingModeSummary: {
            BillingMode: 'PROVISIONED',
            // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
            foo: 'bar',
            baz: undefined,
            // should be ignored
            LastUpdateToPayPerRequestDateTime: new Date(),
          },
        };
        const expected: TableComparisonProperties = {
          BillingModeSummary: {
            BillingMode: 'PAY_PER_REQUEST',
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          BillingModeSummary does not match the expected value.
          Actual: {\\"BillingMode\\":\\"PROVISIONED\\"}
          Expected: {\\"BillingMode\\":\\"PAY_PER_REQUEST\\"}"
        `);
      });

      test('ProvisionedThroughput', () => {
        const actual: TableDescription = {
          ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 5,
            // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
            foo: 'bar',
            baz: undefined,
            // should be ingnored
            LastDecreaseDateTime: new Date(),
            LastIncreaseDateTime: new Date(),
            NumberOfDecreasesToday: 0,
          },
        };
        const expected: TableComparisonProperties = {
          ProvisionedThroughput: {
            ReadCapacityUnits: 5,
            WriteCapacityUnits: 10,
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          ProvisionedThroughput does not match the expected value.
          Actual: {\\"ReadCapacityUnits\\":5,\\"WriteCapacityUnits\\":5}
          Expected: {\\"ReadCapacityUnits\\":5,\\"WriteCapacityUnits\\":10}"
        `);
      });

      test('StreamSpecification', () => {
        const actual: TableDescription = {
          StreamSpecification: {
            StreamEnabled: true,
            StreamViewType: 'NEW_AND_OLD_IMAGES',
            // @ts-expect-error Attribute should be ignored in comparison. AWS SDK may add additional fields in the future
            foo: 'bar',
            baz: undefined,
          },
        };
        const expected: TableComparisonProperties = {
          StreamSpecification: {
            StreamEnabled: false,
            StreamViewType: 'NEW_AND_OLD_IMAGES',
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          StreamSpecification does not match the expected value.
          Actual: {\\"StreamEnabled\\":true,\\"StreamViewType\\":\\"NEW_AND_OLD_IMAGES\\"}
          Expected: {\\"StreamEnabled\\":false,\\"StreamViewType\\":\\"NEW_AND_OLD_IMAGES\\"}"
        `);
      });

      test('SSEDescription', () => {
        const actual: TableDescription = {
          SSEDescription: undefined,
        };
        const expected: TableComparisonProperties = {
          SSEDescription: {
            SSEType: 'KMS',
          },
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          SSEDescription does not match the expected value.
          Actual: undefined
          Expected: {\\"SSEType\\":\\"KMS\\"}"
        `);
      });

      test('DeletionProtectionEnabled', () => {
        const actual: TableDescription = {
          DeletionProtectionEnabled: true,
        };
        const expected: TableComparisonProperties = {
          DeletionProtectionEnabled: false,
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          DeletionProtectionEnabled does not match the expected value.
          Actual: true
          Expected: false"
        `);
      });

      test('multiple errors', () => {
        const actual: TableDescription = {
          SSEDescription: undefined,
          DeletionProtectionEnabled: true,
        };
        const expected: TableComparisonProperties = {
          SSEDescription: {
            SSEType: 'KMS',
          },
          DeletionProtectionEnabled: false,
        };

        expect(() => validateImportedTableProperties(getImportedTableComparisonProperties(actual), expected))
          .toThrowErrorMatchingInlineSnapshot(`
          "Imported table properties did not match the expected table properties.
          SSEDescription does not match the expected value.
          Actual: undefined
          Expected: {\\"SSEType\\":\\"KMS\\"}
          DeletionProtectionEnabled does not match the expected value.
          Actual: true
          Expected: false"
        `);
      });
    });
  });
});
