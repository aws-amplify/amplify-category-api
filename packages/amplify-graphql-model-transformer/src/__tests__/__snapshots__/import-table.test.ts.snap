// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import-table getExpectedTableProperties 1`] = `
Object {
  "AttributeDefinitions": Array [
    Object {
      "AttributeName": "todoId",
      "AttributeType": "S",
    },
    Object {
      "AttributeName": "name",
      "AttributeType": "S",
    },
    Object {
      "AttributeName": "name2",
      "AttributeType": "S",
    },
  ],
  "BillingModeSummary": Object {
    "BillingMode": "PROVISIONED",
  },
  "DeletionProtectionEnabled": false,
  "GlobalSecondaryIndexes": Array [
    Object {
      "IndexName": "byName2",
      "KeySchema": Array [
        Object {
          "AttributeName": "name2",
          "KeyType": "HASH",
        },
      ],
      "Projection": Object {
        "ProjectionType": "ALL",
      },
      "ProvisionedThroughput": Object {
        "ReadCapacityUnits": 5,
        "WriteCapacityUnits": 5,
      },
    },
  ],
  "KeySchema": Array [
    Object {
      "AttributeName": "todoId",
      "KeyType": "HASH",
    },
    Object {
      "AttributeName": "name",
      "KeyType": "RANGE",
    },
  ],
  "ProvisionedThroughput": Object {
    "ReadCapacityUnits": 5,
    "WriteCapacityUnits": 5,
  },
  "SSEDescription": Object {
    "SSEType": "KMS",
    "Status": "ENABLED",
  },
  "StreamSpecification": Object {
    "StreamEnabled": true,
    "StreamViewType": "NEW_AND_OLD_IMAGES",
  },
}
`;
